### ISSUE TITLE
allow set of timeout when solving expressions

### ISSUE THREAD

**Issue by pointhi** [sympy/sympy#9691]
I'm currently developing a plugin to allow sympy support inside searx: asciimoo/searx#384.

The Problem is that this is a web application, and queries like ```solve(Eq(cos(x)^2+sin(2*x+pi),tan(1)),x)``` are able to DDOS searx instances. To solve this we need a clear way to set a a timeout for calculations.

For example, SymPy Gamma has the same issue: http://gamma.sympy.org/input/?i=solve%28Eq%28-sin%282*x%29+%2B+cos%28x%29**2%2Ctan%281%29%29%2C+x%29

**Comment by hargup** [comment]
> I'm currently developing a plugin to allow sympy support inside searx: asciimoo/searx#384.

Great!

> The Problem is that this is a web application, and queries like ```solve(Eq(cos(x)^2+sin(2*x+pi),tan(1)),x)``` are able to DDOS searx instances. To solve this we need a clear way to set a a timeout for calculations.

(nitpick) In python the power operation is represented by the `**` operator and `^` is for bitwise xor operator [1]. We are writing a new `solveset` module to replace `solve` which is clearner and more robust than `solve` and also features like representation of infinite solutions. I invite you to try it out from the development branch. Then I'm not sure if we want to have a timeout functionality in `solveset` and if they will be useful beyond your use case. Since you need it you should go out and try to implement it yourself. Send a PR and we will help you out.

[1]: https://wiki.python.org/moin/BitwiseOperators

**Comment by hargup** [comment]
Plus I'm not sure how can we implement Timeouts, so you should look for solutions which doesn't require modifyications in SymPy code.

**Comment by bjodah** [comment]
Pretty much any non-trivial operation in SymPy faces the risk of "hanging".
Introducing `timeout` kwargs all over the API would soon become very impractical.
So I second the recommendation to solve this outside SymPy. (if the solution is general enough it might be worth providing some kind of convenience function in sympy utilities.)

**Comment by pointhi** [comment]
Ok, to find a sollution, I have to know some of the internal behaviour of sympy:

* is sympy using threads
* is sympy using a pipe, locking ressources or doing other things which cause problems when killing it abrupt.
* is it possible to exit sympy cleanly by raising a exception, and is sympy using system calls (because then exceptions are ignored)

https://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread-in-python

**Comment by debugger22** [comment]
No, SymPy doesn't use threads or pipes and you can exit it by raising an
exception.


Sudhanshu Mishra

**Comment by asmeurer** [comment]
You can implement your own simple timeouts using signal.alarm module. That can still get stuck in functions that are written in C (i.e., builtins), so to be completely safe you may have to implement something outside of the Python process.

**Comment by VahanArsenyan** [comment]
U can use multiprocessing there is a function called get which can get timeout property. Just wrap ur code to that function.

**Comment by Aypac** [comment]
> 
> U can use multiprocessing there is a function called get which can get timeout property. Just wrap ur code to that function.

Yes, a working solution is outlined [on stackoverflow](https://stackoverflow.com/questions/27588919/how-to-call-a-system-command-with-specified-time-limit-in-python).
I still think this could be an easy but valuable addition to sympy, as many operations can virtually take forever, while most of the times, the short operations are the ones that lead to success.

**Comment by gutow** [comment]
I actually favor adding timeouts to things that can potentially run long. This is something that I think could be implemented incrementally. If an operation has a timeout added to it then you add acceptance of the keyword `timeout` to that operation. It appears to me that using any of the other timeout tools well requires running operations in separate threads, which leads to problems with setting up the environment properly. Additionally Windows, which is rather too common, does not appear to play well with python thread implementations.

To do this SymPy needs to define a preferred timeout length, a `SymPyTimeOutError` , and a consistent message on how to override the timeout.

**Comment by asmeurer** [comment]
> It appears to me that using any of the other timeout tools well requires running operations in separate threads, which leads to problems with setting up the environment properly.

I don't think that's the case. You can use signals to get a timeout without using threads, for example. 

But even so, I don't see how implementing the timeout in SymPy itself would change the nature of it. I don't really see the benefit of having

```
function(..., timeout=100)
```

vs.

```
with timeout(100):
    function(...)
```

where `timeout` is some utility like the ones in the stackoverflow answers. The benefit of the later is it works with any function, and doesn't require cluttering the SymPy API.

**Comment by gutow** [comment]
Unfortunately, as I understand it `signals` only works on what I will obnoxiously call ROSes (Real Operating Systems) not Windows. Additionally, timeout wrappers not based on threads can also have their errors trapped by `try...except` in the called functions. See [this](https://stackoverflow.com/questions/492519/timeout-on-a-function-call) often often updated exchange on stackexchange.

Timeouts implemented in Sympy would require the long running operations to check the time passed regularly (every x "cycles"). This would put extra load on the operation, but avoid issues of control and likely be more portable between OSes.

**Comment by oscarbenjamin** [comment]
@gutow do you know of a generic way to implement timeouts on all operating systems including Windows?

If the answer is that timeout-handling code has to be written into every possibly slow part of sympy then that seems a non-starter to me right now. I wouldn't oppose someone else working on that but I certainly wouldn't work on it myself. If there is a clear way to implement a reliable timeout that works on all OSes then maybe we could include a utility to make that accessible to users. I suspect that the fact that this doesn't already exist in the stdlib is because it's not very simple...

**Comment by asmeurer** [comment]
That idea would only work for a very select set of circumstances. Very often when an operation is slow it gets stuck some code that is outside of SymPy, like a list operation. In some cases, code is slow will be slow in different places. This might work for certain algorithms that do involve long loops that iterate many times, for instance, `factorint`. But for example, I don't think your suggested idea would be tenable for something like generic slowdowns in the polys.

**Comment by gutow** [comment]
>...do you know of a generic way to implement timeouts on all operating systems including Windows?

Unfortunately, I have not come up with nor encountered any sneaky/programmer time efficient way to do this for all OSes. The only way I am aware of to do this is to make each potentially long running block of code have its own internal timeout, based on periodically checking the time that has passed. As @asmeurer points out, if the code uses an external operation that runs long that poses additional problems.

I am not very sophisticated with the python tools for checking where code is spending time. But if somebody understands how to do this well and could identify the code blocks most likely to run long, it would inform a discussion of how sophisticated a timeout system would have to be. If the worst holdups are in iterative processes within SymPy, then I believe adding an effective timeout would not be too hard. The results might also suggest code to target for the first try.

**Comment by bjodah** [comment]
We should probably add an example to SymPy's documentation showing how [stopit](https://pypi.org/project/stopit/) (which supports both *nix and Windows) can be used with SymPy:
```python
import stopit
# ...
with stopit.ThreadingTimeout(10) as tt:
    result = sympy.solve(my_eq, x)
```
